awswrangler/lakeformation/_read.py

```diff
@@ -0,0 +1,192 @@
+from pyarrow import NativeFile, RecordBatchStreamReader
+
+from awswrangler import _utils, exceptions
+from awswrangler._config import apply_configs
+from awswrangler.lakeformation._utils import wait_query
+
+_logger: logging.Logger = logging.getLogger(__name__)
+
+
+def _execute_query(
+    query_id: str,
+    token_work_unit: Tuple[str, int],
+    boto3_session: Optional[boto3.Session] = None,
+) -> pd.DataFrame:
+    client_lakeformation: boto3.client = _utils.client(service_name="lakeformation", session=boto3_session)
+    token, work_unit = token_work_unit
+    messages: NativeFile = client_lakeformation.execute(QueryId=query_id, Token=token, WorkUnitId=work_unit)["Messages"]
+    return RecordBatchStreamReader(messages.read()).read_pandas()
+
+
+def _resolve_sql_query(
+    query_id: str,
+    chunked: Optional[bool] = None,
+    use_threads: bool = True,
+    boto3_session: Optional[boto3.Session] = None,
+) -> Union[pd.DataFrame, Iterator[pd.DataFrame]]:
+    client_lakeformation: boto3.client = _utils.client(service_name="lakeformation", session=boto3_session)
+
+    wait_query(query_id=query_id, boto3_session=boto3_session)
+
+    # The LF Query Engine distributes the load across workers
+    # Retrieve the tokens and their associated work units until NextToken is ''
+    # One Token can span multiple work units
+    # PageSize determines the size of the "Units" array in each call
+    # TODO: Inquire about good page size # pylint: disable=W0511
+    scan_kwargs: Dict[str, Union[str, int]] = {"QueryId": query_id, "PageSize": 2}
+    next_token: str = "init_token"  # Dummy token
+    token_work_units: List[Tuple[str, int]] = []
+    while next_token:
+        response = client_lakeformation.get_work_units(**scan_kwargs)
+        token_work_units.extend(  # [(Token0, WorkUnitId0), (Token0, WorkUnitId1), (Token1, WorkUnitId2) ... ]
+            [
+                (unit["Token"], unit_id)
+                for unit in response["Units"]
+                for unit_id in range(unit["WorkUnitIdMin"], unit["WorkUnitIdMax"] + 1)  # Max is inclusive
+            ]
+        )
+        next_token = response["NextToken"]
+        scan_kwargs["NextToken"] = next_token
+
+    dfs: List[pd.DataFrame] = list()
+    if use_threads is False:
+        dfs = list(
+            _execute_query(query_id=query_id, token_work_unit=token_work_unit, boto3_session=boto3_session)
+            for token_work_unit in token_work_units
+        )
+    else:
+        cpus: int = _utils.ensure_cpu_count(use_threads=use_threads)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=cpus) as executor:
+            dfs = list(
+                executor.map(
+                    _execute_query,
+                    itertools.repeat(query_id),
+                    token_work_units,
+                    itertools.repeat(_utils.boto3_to_primitives(boto3_session=boto3_session)),
+                )
+            )
+    dfs = [df for df in dfs if not df.empty]
```

Comment:
Required because the `execute` function can return empty pd.DataFrames

Reply:
Seems good to me.